### ISSUE TITLE
takeUntilDestroyed might not work as expected in certain situations

### ISSUE THREAD

**Issue by yjaaidi** [angular/angular#49658]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

The current implementation of `takeUntilDestroyed()` operator assumes that the operator is used before the component is destroyed.
If the operator is used after the component is destroyed then it will not unsubscribe as expected and observables might leak.

Here is an example of misuse where this can happen:

```ts
class MyCmp {
 destroyRef = inject(DestroyRef);
 ngOnDestroy() {
  setTimeout(() => source$.pipe(takeUntilDestroyed(this.destroyRef).subscribe());
 }
}
```

While in most cases, this wouldn't happen as the injection context would be missing... but then we can pass a `destroyRef` manually like in the example above.

Cf. https://github.com/angular/angular/blob/2703fd626040c5e65401ebd776404a3b9e284724/packages/core/rxjs-interop/src/take_until_destroyed.ts#L29-L31

It might be interesting to add a `isDestroyed` property on `DestroyRef` for late subscriber use cases like `takeUntilDestroyed()` usage above.

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular 16.0.0-next.5
```

**Comment by pkozlowski-opensource** [comment]
An alternative would be to immediately invoke the onDestroy callback when the DestroyRef is out of its lifecycle scope.

**Comment by yjaaidi** [comment]
**TL;DR:** 💯 for `DestroyRef.onDestroy()` to throw an error if already destroyed.

> We would rather throw an informative error in that case. 

This is definitely my favorite solution. I really like it as it encourages & inspires better design.

> What would be the usage of `isDestroyed`?

To throw an error like described above. Of course, it's better if it's handled in `DestroyRef`. 
I didn't consider as an option initially as I didn't know the implications (i.e. existing usages of `DestroyRef`).

> An alternative would be to immediately invoke the onDestroy callback when the DestroyRef is out of its lifecycle scope.
While this would work, I still prefer the option of throwing an error for multiple reasons:
- pedagogical reasons as it teaches better design and could link to a resource explaining RxJS flattening strategies etc...
- invoking the `onDestroy` callback immediately if already destroyed is kind of surprising. Its name indicates an event and the late subscriber issue shouldn't be handled at this level.
- in some "extreme situations", having the callback be triggered both synchronously or asynchronously can cause trouble.
e.g. 
```ts
let name = 'foo';
destroyRef.onDestroy(() => name = null)
console.log(user); // null or not depending if the callback is executed synchronously or not.
``` 
Of course, we shouldn't rely on that... but eventually, someone will.

**Comment by yjaaidi** [comment]
Thanks @pkozlowski-opensource ! That was quick! Good job!
